textgrad/engine/bedrock.py

```diff
@@ -0,0 +1,157 @@
+                if self.aws_region is None:
+                    raise ValueError("AWS region not specified. Please add it in get_engine parameters or has AWS_DEFAULT_REGION var")
+            if access_key_id is None:
+                raise ValueError("AWS access key ID cannot be 'None'.")
+            if secret_access_key is None:
+                raise ValueError("AWS secret access key cannot be 'None'.")
+            session = boto3.Session(
+                aws_access_key_id=access_key_id,
+                aws_secret_access_key=secret_access_key,
+                aws_session_token=session_token
+            )
+            self.my_config = Config(region_name = self.aws_region)
+            self.client = session.client(service_name='bedrock-runtime', config=self.my_config)
+
+        root = platformdirs.user_cache_dir("textgrad")
+        cache_path = os.path.join(root, f"cache_bedrock_{model_string}.db")
+        super().__init__(cache_path=cache_path)
+        
+        self.model_string = model_string
+        self.system_prompt = system_prompt
+        
+        assert isinstance(self.system_prompt, str)
+
+    @retry(wait=wait_random_exponential(min=1, max=5), stop=stop_after_attempt(5))
+    def __call__(self, prompt, **kwargs):
+        return self.generate(prompt, **kwargs)
+    
+    def generate_conversation(self, model_id="", system_prompts=[], messages=[], temperature=0.5, top_k=200, top_p=0.99, max_tokens=2048):
+        """
+        Sends messages to a model.
+        Args:
+            bedrock_client: The Boto3 Bedrock runtime client.
+            model_id (str): The model ID to use.
+            system_prompts (JSON) : The system prompts for the model to use.
+            messages (JSON) : The messages to send to the model.
+
+        Returns:
+            response (JSON): The conversation that the model generated.
+
+        """
+
+        # Base inference parameters to use.
+        inference_config = {"temperature": temperature, "topP": top_p, "maxTokens": self.max_tokens if self.max_tokens else max_tokens}
+        if("anthropic" in model_id): 
+            # Additional inference parameters to use.
+            additional_model_fields = {"top_k": top_k}
+        else: 
+            additional_model_fields = {}
+        
+        # Send the message.
+        if self.system_prompt_supported:
+            response = self.client.converse(
+                modelId=model_id,
+                messages=messages,
+                system=system_prompts,
+                inferenceConfig=inference_config,
+                additionalModelRequestFields=additional_model_fields
+            )
+        else:
+            response = self.client.converse(
+                modelId=model_id,
+                messages=messages,
+                inferenceConfig=inference_config,
+                additionalModelRequestFields=additional_model_fields
+            )
+
+        return response
+
+    def generate(
```

Comment:
this is great! Do you want to add support for multiple inputs to this engine? Like [here](https://github.com/zou-group/textgrad/blob/main/textgrad/engine/openai.py#L116)?

Reply:
yes I can do! Can you provide me a simple example to test the implementation out?

Regarding to images only Anthropic models in Bedrock support vision so I'll enable that only for them!